File: Python/Stacks & Queues/stack.py
# Problem link - https://www.tutorialspoint.com/javaexamples/data_stack.htm


class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class Stack:
    def __init__(self):
        self.head = self.tail = None
        self.length = 0

    def is_empty(self):
        return self.length == 0

    def top(self):
        if self.is_empty():
            return
        return self.head.data

    def push(self, x):
        node = Node(x)
        if self.is_empty():
            self.head = self.tail = node
        else:
            node.next = self.head
            self.head = node
        self.length += 1

    def pop(self):
        if self.is_empty():
            return
        node = self.head
        data = self.head.data
        self.head = self.head.next
        del node
        self.length -= 1
        return data


# Example 1
stack = Stack()
for i in [1, 2, 3, 4]:
    stack.push(i)
while not stack.is_empty():
    print(stack.pop(), end=" ")
print()

File: Java/StacksAndQueues/Problem12/Stack.java
package StacksAndQueues.Problem12;


class Node<T> {
    public T data;
    public Node<T> next;

    public Node(T data) {
        this.data = data;
        this.next = null;
    }
}


public class Stack<T> {
    public Node<T> head;
    public Node<T> tail;
    public Integer length;

    public Stack() {
        this.head = this.tail = null;
        this.length = 0;
    }

    public boolean isEmpty() {
        return this.length == 0;
    }

    public void push(T x) {
        Node<T> node = new Node<>(x);
        if (isEmpty()) {
            head = tail = node;
        } else {
            node.next = head;
            head = node;
        }
        length += 1;
    }

    public T pop() {
        if (isEmpty()) return null;
        T item = head.data;
        head = head.next;
        length -= 1;
        return item;
    }

    public T top() {
        if (isEmpty()) return null;
        return head.data;
    }
}
